File: lab03/z1_c/parrot.c
#include <stdlib.h>
#include <stdio.h>

typedef char const *(*PTRFUN)();

typedef struct {
    PTRFUN *vtable;
    char const *name;
} Parrot;

char const* name(void* this) {
    return ((Parrot*)this)->name;
}
char const* greet() {
    return "Sqwak!";
}
char const* menu() {
    return "sjemenke";
}

PTRFUN functions[3] = {name, greet, menu};

void* create(char const *name) {
    Parrot *parrot = (Parrot*)malloc(sizeof(Parrot));
    parrot->vtable = functions;
    parrot->name = name;
    return parrot;
}

File: lab03/lab3_java/src/lab3/z1/model/plugins/Parrot.java
package lab3.z1.model.plugins;

import lab3.z1.model.Animal;

public class Parrot extends Animal {

	private String name;
	
	public Parrot(String name) {
		this.name = name;
	}
	
	@Override
	public String name() {
		return name;
	}

	@Override
	public String greet() {
		return "sqwak!";
	}

	@Override
	public String menu() {
		return "sjemenke";
	}

}
